### DISCUSSION TITLE
Cask 'x2goclient' definition is invalid: invalid depends_on key: ':x11'

### DISCUSSION THREAD

**Discussion by kevdogg** [Homebrew/discussions#2621] [Casks]
I'm aware that Mac homebrew is trying to migrate away from X11 (Xquartz) dependencies and rely on their own x11 libraries.  I don't believe the x2goclient formula has been updated to reflect these changes.  

I read about the changes via this link: https://github.com/Homebrew/homebrew-core/issues/64166 however I when stepping threw some of the steps listed I couldn't generate any linkage.  

Does this X11 change a thing the end user of the formulae needs to reconcile or is this more of on the formulae developer?

**Comment by gromgit** [comment]
> I don't believe the x2goclient formula has been updated to reflect these changes.

1. AFAICT, there's never been an `x2goclient` _formula_ in Homebrew core, i.e. a "script" to, among other things, build from source.
2. There is an `x2goclient` cask definition, but that lost its X11 dependency almost a year ago as of this writing.

If you're actually talking about the `x2goclient` Homebrew-hosted cask, your Homebrew cask repo is _seriously_ out of date. Also, the stuff mentioned in https://github.com/Homebrew/homebrew-core/issues/64166 has no relevance to casks; since they're binaries compiled and delivered by third parties, no incantation on the Homebrew side can magically cause said binaries to suddenly sprout X11 linkages.

But if you're talking about an actual formula, then you're looking at someone else's repo, not Homebrew core, in which case they're probably the ones you should take this up with.

**Reply by kevdogg** [reply]
Clearly I'm using the wrong terminology.  I'm speaking of the x2goclient cask directly.  

```
#brew reinstall --cask x2goclient                         
==> Downloading https://code.x2go.org/releases/X2GoClient_latest_macosx_10_13.dmg
Already downloaded: /Users/kevdog/Library/Caches/Homebrew/downloads/926bc503da4d0c456f58ef590697309e9f356799cc5f463d3b33bb2e000787e7--X2GoClient_latest_macosx_10_13.dmg
Warning: No checksum defined for cask 'x2goclient', skipping verification.
Error: Cask 'x2goclient' definition is invalid: invalid depends_on key: ':x11'
```

I don't think my repositories are out of date??

```
#brew tap                                                       
buo/cask-upgrade
gromgit/fuse
homebrew/bundle
homebrew/cask
homebrew/cask-drivers
homebrew/cask-fonts
homebrew/cask-versions
homebrew/core
teamookla/speedtest

```

```
#brew outdated                                                
x2goclient (4.1.2.2,20200213) != 4.1.2.2
```

**Reply by SMillerDev** [reply]
do `brew config` to see the age of your taps and `brew update` to update them

`brew tap` only shows which taps you have
`brew outdated` only shows outdated formula

**Comment by kevdogg** [comment]
I tried updating prior to upgrading but still received the x11 error message.  

```
# brew config                                                    
HOMEBREW_VERSION: 3.3.7-38-g6295b08
ORIGIN: https://github.com/Homebrew/brew
HEAD: 6295b082a970c5036185b6f997632f009225519b
Last commit: 16 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 9dcb182af3e5aab9ee486c2475323ff527465d9b
Core tap last commit: 11 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_CORE_GIT_REMOTE: https://github.com/Homebrew/homebrew-core
HOMEBREW_DISPLAY: :0
HOMEBREW_EDITOR: mvim
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.8 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.8/bin/ruby
CPU: octa-core 64-bit haswell
Clang: 13.0.0 build 1300
Git: 2.34.1 => /usr/local/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 11.6.1-x86_64
CLT: 13.0.0.0.1.1630607135
Xcode: 13.1
```

**Reply by gromgit** [reply]
What's the output of `brew log --cask x2goclient | head`? If it shows a datestamp older than Nov 2 2021, your cask repo is out-of-date:
```
$ brew log --cask x2goclient | head
commit 31e86bd2986c51a5e50b1338fa21112e1b11d68c
Author: neutric <5984479+neutric@users.noreply.github.com>
Date:   Tue Nov 2 06:46:11 2021 +0100

    x2goclient: add desc (#113619)
    
    * x2goclient: add desc
    
    * Update x2goclient.rb
    
```

**Comment by kevdogg** [comment]
```
# brew log --cask x2goclient | head                              
commit 31e86bd2986c51a5e50b1338fa21112e1b11d68c
Author: neutric <5984479+neutric@users.noreply.github.com>
Date:   Tue Nov 2 06:46:11 2021 +0100

    x2goclient: add desc (#113619)

    * x2goclient: add desc

    * Update x2goclient.rb
```

**Reply by gromgit** [reply]
And does `git -C $(brew --repo homebrew/cask) diff Casks/x2goclient.rb` output anything?

**Reply by kevdogg** [reply]
I ran that command -- blank output. 

```
git -C $(brew --repo homebrew/cask) diff Casks/x2goclient.rb
(END)
```

**Reply by gromgit** [reply]
I just noticed that you're trying to _reinstall_ the cask. Try instead:
```
brew rm --cask x2goclient
brew install --cask x2goclient
```
and see if you still hit the same X11 dependency error.

**Reply by kevdogg** [reply]
Definitely having a bit of an issue here  -- any other way to force remove:

```
# brew rm --cask --force --ignore-dependencies x2goclient       
Error: Cask 'x2goclient' definition is invalid: invalid depends_on key: ':x11'
```

**Reply by gromgit** [reply]
Looks like you'll have to nuke all traces of the old cask definition first:
```
rm -fr $(brew --prefix)/Caskroom/x2goclient
brew rm --cask --force x2goclient
brew install --cask x2goclient
```

**Reply by kevdogg** [reply]
Thanks those instructions helped.  I have no idea why my install bugged out, but your instructions did the trick!!!

**Reply by gromgit** [reply]
I'm not a Homebrew maintainer, but I suspect that for casks, the original definition needs to be kept around for cleanup purposes. And when that old definition contains syntax that's no longer supported...
